### DISCUSSION TITLE
Wagtail Oderable Sort Order Incorrectly Set

### DISCUSSION THREAD

**Discussion by valentinemthembu** [wagtail/wagtail#9249] [General]
We have inline panels using ParentalKey , The Parent models is Form and the child model is FormFields.
When we create the Form for the first time, the FormFields are arranged in a particular order (Image 1). The issue we are facing is that when the Form is saved,   the sort_order is assigned based on how the fields are saved to the database (image 2: primary key). Which then makes the ordering incorrect. When we save the Form again after rearranging the FormFields the order at this point is save correctly.

Payload when the form is created.
{"pk": null, "banner_text": "Form Testing", "form_site": 3, "heading": "Form Testing", "internal_name": "Form Testing", "requires_auth": false, "button_text": "Form Testing", "classification": 5, "icon": 84, "banner_color": "#e1f7d5", "form_fields": []}

Payload when the form is updated.
{"pk": 153, "banner_text": "logging.warning([self.to](http://self.to/)_json())", "form_site": 3, "heading": "logging.warning([self.to](http://self.to/)_json())", "internal_name": "logging.warning([self.to](http://self.to/)_json())", "requires_auth": false, "button_text": "logging.warning([self.to](http://self.to/)_json())", "classification": 5, "icon": 84, "banner_color": "#e1f7d5", "form_fields": [{"pk": 682, "sort_order": 0, "form": 153, "field": 4, "is_required": false}, {"pk": 681, "sort_order": 1, "form": 153, "field": 3, "is_required": false}]}

How the save works.
Form is save first and assigned a primary key
After Form is saved, then FormFields are lopped through and saved.
This is when the issue is experienced, When the fields are being looped through, they are passed into the save method incorrectly.
Looking at the inline panel class, when we click the add button it opens a transactions that is only committed when we save the form.

**Comment by lb-** [comment]
Pretty sure this is a duplicate of a known bug in 4.0. 

https://github.com/wagtail/wagtail/issues/9145

Fix coming soon (this week hopefully).

**Reply by valentinemthembu** [reply]
Thank you for the response. I have taken a look at the issue above. I will revist after the fix goes in.

**Comment by lb-** [comment]
Locking this conversation as this is fixed in Wagtail 4.0.2
